### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by beanhuo**
This series patch is to add support for unipro sub-layers attributes read/write.

### CODE REVIEW

**Review by arthur-simchaev-sndk — APPROVED** [review]
Good

**Review by arthur-simchaev-sndk — APPROVED** [review]
Good

**Review by arthur-simchaev-sndk — APPROVED** [review]

File: scsi_bsg_util.c [hunk 1]
```diff
@@ -111,8 +111,7 @@ int send_bsg_sg_io(int fd, struct ufs_bsg_request *request_buff,
 #endif
 
 	while (((ret = ioctl(fd, SG_IO, &io_hdr_v4)) < 0) &&
-		((errno == EINTR) || (errno == EAGAIN)))
-		;
+		((errno == EINTR) || (errno == EAGAIN)));
```
Comment (arthur-simchaev-sndk) [inline scsi_bsg_util.c]: Avri , please comment this change . Since the code was change due to your comment

File: options.c [hunk 2]
```diff
@@ -80,6 +95,19 @@ int init_options(int opt_cnt, char *opt_arr[], struct tool_options *options)
 	return rc;
 }
 
+static int verify_target(struct tool_options *options, int target)
+{
+	if (options->target != INVALID) {
+		print_error("duplicated operate target.");
+		goto out;
+	}
+    options->target = target;
```
Comment (arthur-simchaev-sndk) [inline options.c]: Please replace spaces with tab

File: unipro.c [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+	{8, "PEER_COMMUNICATION_FAILURE"},
+	{9, "BUSY"},
+	{10, "DME_FAILURE"},
+};
+
+int ufshcd_dme_get_attr(int fd, __u32 attr_sel, __u8 peer)
+{
+	struct ufs_bsg_request bsg_req = { 0 };
+	struct ufs_bsg_reply bsg_rsp = { 0 };
+	struct uic_command *uic_cmd = (struct uic_command *)&bsg_req.upiu_req.uc;
```
Comment (arthur-simchaev-sndk) [inline unipro.c]: General comments : The code was written according to Linux Kernel Coding standard. 
Could you please run checkpatch.pl & fix all warnings . 
E.g.  line over 80 characters,  unnecessary whitespace before etc...

File: unipro.c [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+	default:
+		rt = INVALID;
+		break;
+	}
+	close(fd);
+	return rt;
+}
+
+void unipro_help(char *tool_name)
+{
```
Comment (arthur-simchaev-sndk) [inline unipro.c]: General comments . Could you please clarify which test did you run m, in order to verify UIS feature?
Thnx

File: unipro.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Comment (arthur-simchaev-sndk) [inline unipro.h]: Please use /* comment instead of //  in license identifier

File: unipro.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+
+struct ufs_uic_attr_fields {
+	const char *name;
+	__u32 id;
+	enum unipro_acc_mode acc_mode;
+};
+
+struct ufs_unipro_attrs_info {
+
```
Comment (arthur-simchaev-sndk) [inline unipro.h]: Remove redundant line

File: unipro.c [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+	    if (res_code < (max_code - 1))
+		    print_error("%s: attr-id 0x%x %s.\n", __func__,
+				UIC_GET_ATTR_ID(attr_sel), UIC_ConfigResultCode[res_code].def);
+	    else
+		    print_error("%s: attr-id 0x%x, fail to identify error code %d.\n", __func__,
+				UIC_GET_ATTR_ID(attr_sel), res_code);
+
+		rt = ERROR;
+	} else {
+		rt = uic_rsq.argument3;
```
Comment (arthur-simchaev-sndk) [inline unipro.c]: Remove redundant brackets

File: unipro.c [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+}
+
+static int unipro_read(int fd, int idn, int id, __u8 all)
+{
+
+	int index, qts;
+	int mib_val_local, mib_val_peer;
+	int ret = OK;
+
+	if (idn >= MAX_UNIPRO_IDN) {
```
Comment (arthur-simchaev-sndk) [inline unipro.c]: Please move it to option file,when the tool verify the parameters

File: scsi_bsg_util.c (hunk 1 above)
Comment (avri-altman-sndk) [reply scsi_bsg_util.c]: Please note that this is not a do {} while (); statement.
We emphasizes the semicolon, making it clear that the body of the while is an empty statement.

for more info why we should do that please look for "-ss" in https://www.gnu.org/software/indent/manual/indent.html

File: options.c [hunk 9]
```diff
@@ -92,7 +120,11 @@ static int verify_and_set_index(struct tool_options *options)
 	/* In case atoi returned 0 . Check that is real 0 and not error
 	 * arguments . Also check that the value is in correct range
 	 */
-	index = atoi(optarg);
```
Comment (avri-altman-sndk) [inline options.c]: this should go in a separate [BUGFIX] patch

File: ufs.c [hunk 10]
```diff
@@ -64,6 +66,7 @@ static void initialized_options(struct tool_options *options)
 	memset(options, INVALID, sizeof(*options));
 	options->path[0] = '\0';
 	options->data = NULL;
+
```
Comment (avri-altman-sndk) [inline ufs.c]: please do not mix coding style fixes with functional fixes.

File: options.c [hunk 11]
```diff
@@ -9,6 +9,7 @@
 
 #include "options.h"
 #include "ufs.h"
+#include "unipro.h"
```
Comment (avri-altman-sndk) [inline options.c]: this patch will not compile without unipro.h

File: unipro.c (hunk 7 above)
Comment (avri-altman-sndk) [reply unipro.c]: actually those brackets should stay:
if () {
...
} else {
...
}

File: unipro.c [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+out:
+	return rt;
+}
+
+int check_attr_id(__u32 idn, __u32 id)
+{
+	int index;
+	int qts = uic_attrs_group[idn].items;
+	struct ufs_uic_attr_fields *p = uic_attrs_group[idn].attrs;
+	for (index = 0; index < qts; index++) {
```
Comment (avri-altman-sndk) [inline unipro.c]: one line space please

File: ufs.c [hunk 13]
```diff
@@ -13,6 +13,7 @@
 #include "ufs_cmds.h"
 #include "options.h"
 #include "ufs.h"
+#include "unipro.h"
```
Comment (avri-altman-sndk) [inline ufs.c]: you forgot to add unipro.h

File: options.c [hunk 14]
```diff
@@ -80,6 +95,19 @@ int init_options(int opt_cnt, char *opt_arr[], struct tool_options *options)
 	return rc;
 }
 
+static int verify_target(struct tool_options *options, int target)
```
Comment (avri-altman-sndk) [inline options.c]: in case of read, you don't need to know if its a peer or local - you are reading both.
 I think you should at least make a comment of it.
 
 If unipro.h is already included, you can add to this patch all further checks,
 e.g. of the unipro idn etc.

**Comment by beanhuo** [comment]
OK, let's follow the Linux C coding style. will send next formatted version.
